File: src/lib/services/employee.service.ts
import { employeeRepository, EmployeeCreateData } from "@/lib/repositories/employee.repository"
import { auditRepository } from "@/lib/repositories/dashboard.repository"
import { toDate, toFloat, strOrNull, str } from "@/lib/validation"
import { db } from "@/lib/db"

const EMPLOYEE_FIELDS = new Set([
  "employeeCode", "firstName", "middleName", "lastName", "displayName", "gender",
  "dateOfBirth", "maritalStatus", "bloodGroup", "nationality", "profilePhotoUrl",
  "personalEmail", "officialEmail", "mobileNumber", "alternateNumber",
  "dateOfJoining", "employmentType", "workerType", "probationStatus", "probationEndDate",
  "confirmationDate", "noticePeriod", "employeeStatus",
  "entityId", "branchId", "departmentId", "designationId", "gradeId", "locationId",
  "reportingManagerId",
  "bankName", "accountNumber", "ifscCode", "branchName",
  "panNumber", "aadhaarNumber", "uanNumber", "pfNumber", "esiNumber",
  "currentAddress", "permanentAddress",
  "ctc", "basicSalary", "hra",
])

export interface EmployeeListParams {
  tenantId: string
  q?: string
  departmentId?: string
  entityId?: string
  branchId?: string
  status?: string
  page: number
  pageSize: number
  sortBy?: string
  sortOrder?: "asc" | "desc"
}

export const employeeService = {
  async list(params: EmployeeListParams) {
    const { items, total } = await employeeRepository.findMany({
      tenantId: params.tenantId,
      q: params.q,
      departmentId: params.departmentId,
      entityId: params.entityId,
      branchId: params.branchId,
      status: params.status,
      skip: (params.page - 1) * params.pageSize,
      take: params.pageSize,
      sortBy: params.sortBy,
      sortOrder: params.sortOrder,
    })
    return { items, total, page: params.page, pageSize: params.pageSize }
  },

  async getById(id: string, tenantId: string) {
    const rec = await employeeRepository.findById(id, tenantId)
    if (!rec) return null

    const mgrSelect = { id: true, firstName: true, lastName: true, employeeCode: true, displayName: true } as const
    let reportingManager = null
    let functionalManager = null
    let hrManager = null

    if (rec.reportingManagerId) {
      reportingManager = await db.employee.findFirst({ where: { id: rec.reportingManagerId, tenantId }, select: mgrSelect })
    }
    if (rec.functionalManagerId) {
      functionalManager = await db.employee.findFirst({ where: { id: rec.functionalManagerId, tenantId }, select: mgrSelect })
    }
    if (rec.hrManagerId) {
      hrManager = await db.employee.findFirst({ where: { id: rec.hrManagerId, tenantId }, select: mgrSelect })
    }

    return { ...rec, reportingManager, functionalManager, hrManager }
  },

  async create(body: Record<string, unknown>, tenantId: string, actorName?: string) {
    const employeeCode = str(body.employeeCode).trim()
    const firstName = str(body.firstName).trim()
    if (!employeeCode) throw new ServiceError("Employee code is required", 400)
    if (!firstName) throw new ServiceError("First name is required", 400)

    const existing = await employeeRepository.findByCode(tenantId, employeeCode)
    if (existing) throw new ServiceError("Employee code already exists", 409)

    const extras: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(body)) {
      if (!EMPLOYEE_FIELDS.has(k)) extras[k] = v
    }
    const customData = Object.keys(extras).length ? JSON.stringify(extras) : null

    const data: EmployeeCreateData = {
      tenantId,
      employeeCode,
      firstName,
      middleName: strOrNull(body.middleName),
      lastName: strOrNull(body.lastName),
      displayName: strOrNull(body.displayName),
      gender: strOrNull(body.gender),
      dateOfBirth: toDate(body.dateOfBirth),
      maritalStatus: strOrNull(body.maritalStatus),
      bloodGroup: strOrNull(body.bloodGroup),
      nationality: str(body.nationality, "Indian"),
      personalEmail: strOrNull(body.personalEmail),
      officialEmail: strOrNull(body.officialEmail),
      mobileNumber: strOrNull(body.mobileNumber),
      alternateNumber: strOrNull(body.alternateNumber),
      dateOfJoining: toDate(body.dateOfJoining),
      employmentType: str(body.employmentType, "Full-time"),
      workerType: str(body.workerType, "Permanent"),
      employeeStatus: str(body.employeeStatus, "Active"),
      entityId: strOrNull(body.entityId),
      branchId: strOrNull(body.branchId),
      departmentId: strOrNull(body.departmentId),
      designationId: strOrNull(body.designationId),
      gradeId: strOrNull(body.gradeId),
      locationId: strOrNull(body.locationId),
      reportingManagerId: strOrNull(body.reportingManagerId),
      bankName: strOrNull(body.bankName),
      accountNumber: strOrNull(body.accountNumber),
      ifscCode: strOrNull(body.ifscCode),
      panNumber: strOrNull(body.panNumber),
      aadhaarNumber: strOrNull(body.aadhaarNumber),
      ctc: toFloat(body.ctc),
      basicSalary: toFloat(body.basicSalary),
      hra: toFloat(body.hra),
      customData,
    }

    const rec = await employeeRepository.create(data)

    await auditRepository.log({
      tenantId,
      module: "employees",
      action: "Create",
      recordId: rec.id,
      userName: actorName,
      details: JSON.stringify({ employeeCode, firstName }),
    })

    await db.employeeTimelineEvent.create({
      data: {
        tenantId,
        employeeId: rec.id,
        eventType: "Created",
        title: "Employee record created",
        eventDate: new Date(),
        actorName: actorName || "System",
      },
    })

    return rec
  },

  async update(id: string, body: Record<string, unknown>, tenantId: string, actorName?: string) {
    const existing = await db.employee.findFirst({ where: { id, tenantId } })
    if (!existing) throw new ServiceError("Employee not found", 404)

    if (body.employeeCode !== undefined) {
      const newCode = str(body.employeeCode).trim()
      if (!newCode) throw new ServiceError("Employee code cannot be empty", 400)
      if (newCode !== existing.employeeCode) {
        const clash = await employeeRepository.findByCode(tenantId, newCode)
        if (clash) throw new ServiceError("Employee code already exists", 409)
      }
    }

    const data: Record<string, unknown> = {}
    const changed: { field: string; oldVal: string; newVal: string; module?: string }[] = []

    function setField<T extends string | number | boolean | Date | null>(
      field: string,
      value: T | undefined,
      opts?: { module?: string }
    ) {
      if (value === undefined) return
      const oldVal = (existing as any)[field]
      if (oldVal === value) return
      const normalize = (v: any) => {
        if (v instanceof Date) return v.toISOString()
        if (typeof v === "boolean") return String(v)
        if (v === null || v === undefined) return ""
        return String(v)
      }
      if (normalize(oldVal) === normalize(value)) return
      data[field] = value
      changed.push({
        field,
        oldVal: oldVal === null || oldVal === undefined ? "" : String(oldVal),
        newVal: value === null ? "" : String(value),
        module: opts?.module,
      })
    }

    const fieldMap: [string, () => any, string][] = [
      ["employeeCode", () => str(body.employeeCode).trim(), "Personal"],
      ["firstName", () => str(body.firstName).trim(), "Personal"],
      ["middleName", () => strOrNull(body.middleName), "Personal"],
      ["lastName", () => strOrNull(body.lastName), "Personal"],
      ["displayName", () => strOrNull(body.displayName), "Personal"],
      ["gender", () => strOrNull(body.gender), "Personal"],
      ["dateOfBirth", () => toDate(body.dateOfBirth), "Personal"],
      ["maritalStatus", () => strOrNull(body.maritalStatus), "Personal"],
      ["bloodGroup", () => strOrNull(body.bloodGroup), "Personal"],
      ["nationality", () => str(body.nationality, "Indian"), "Personal"],
      ["dateOfJoining", () => toDate(body.dateOfJoining), "Job"],
      ["employmentType", () => str(body.employmentType, "Full-time"), "Job"],
      ["employeeStatus", () => str(body.employeeStatus, "Active"), "Job"],
      ["departmentId", () => strOrNull(body.departmentId), "Job"],
      ["designationId", () => strOrNull(body.designationId), "Job"],
      ["branchId", () => strOrNull(body.branchId), "Job"],
      ["entityId", () => strOrNull(body.entityId), "Job"],
      ["gradeId", () => strOrNull(body.gradeId), "Job"],
      ["locationId", () => strOrNull(body.locationId), "Job"],
      ["reportingManagerId", () => strOrNull(body.reportingManagerId), "Job"],
      ["ctc", () => toFloat(body.ctc), "Compensation"],
      ["basicSalary", () => toFloat(body.basicSalary), "Compensation"],
      ["hra", () => toFloat(body.hra), "Compensation"],
    ]

    for (const [field, getValue, module] of fieldMap) {
      if (field in body) setField(field, getValue(), { module })
    }

    const extras: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(body)) {
      if (!EMPLOYEE_FIELDS.has(k)) extras[k] = v
    }
    if (Object.keys(extras).length) {
      data.customData = JSON.stringify(extras)
    }

    if (Object.keys(data).length === 0) return existing

    const updated = await employeeRepository.update(id, data)

    if (changed.length) {
      await db.employeeAuditLog.createMany({
        data: changed.map((c) => ({
          tenantId,
          employeeId: id,
          module: c.module || "Personal",
          field: c.field,
          oldValue: c.oldVal,
          newValue: c.newVal,
          action: "Update",
          changedBy: actorName || "HR Admin",
        })),
      })
    }

    const oldStatus = existing.employeeStatus
    const newStatus = data.employeeStatus as string | undefined
    if (newStatus && newStatus !== oldStatus) {
      await db.employeeStatusHistory.create({
        data: {
          tenantId, employeeId: id, oldStatus, newStatus,
          effectiveDate: new Date(),
          changedBy: actorName || "HR Admin",
        },
      })
      await db.employeeTimelineEvent.create({
        data: {
          tenantId, employeeId: id,
          eventType: "Status changed",
          title: `Status changed from "${oldStatus}" to "${newStatus}"`,
          eventDate: new Date(), actorName: actorName || "HR Admin",
        },
      })
    }

    await auditRepository.log({
      tenantId,
      module: "employees",
      action: "Update",
      recordId: id,
      userName: actorName,
      details: JSON.stringify(changed.map((c) => c.field)),
    })

    return updated
  },

  async delete(id: string, tenantId: string, actorName?: string) {
    const existing = await db.employee.findFirst({ where: { id, tenantId } })
    if (!existing) throw new ServiceError("Employee not found", 404)

    await employeeRepository.delete(id)

    await auditRepository.log({
      tenantId,
      module: "employees",
      action: "Delete",
      recordId: id,
      userName: actorName,
      details: JSON.stringify({ employeeCode: existing.employeeCode }),
    })

    return { ok: true }
  },
}

export class ServiceError extends Error {
  status: number
  constructor(message: string, status: number = 400) {
    super(message)
    this.status = status
  }
}

File: backend/app/services/employee_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.repositories.employee_repository import EmployeeRepository
from app.models.models import EmployeeAuditLog, EmployeeTimelineEvent, EmployeeStatusHistory
from app.api.crud_factory import _serialize
from fastapi import HTTPException, status
from datetime import datetime
import json


EMPLOYEE_FIELDS = {
    "employee_code", "first_name", "middle_name", "last_name", "display_name", "gender",
    "date_of_birth", "marital_status", "blood_group", "nationality", "profile_photo_url",
    "personal_email", "official_email", "mobile_number", "alternate_number",
    "date_of_joining", "employment_type", "worker_type", "employee_status",
    "entity_id", "branch_id", "department_id", "designation_id", "grade_id", "location_id",
    "reporting_manager_id", "bank_name", "account_number", "ifsc_code",
    "pan_number", "aadhaar_number", "ctc", "basic_salary", "hra",
}


class EmployeeService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.repo = EmployeeRepository(db)

    def _serialize_with_relations(self, emp) -> dict:
        """Serialize employee to camelCase dict with nested department/designation/entity/location."""
        from app.models.models import Department, Designation, Entity, Location
        result = _serialize(emp)
        # Note: relations are loaded eagerly by the repository
        if hasattr(emp, "_department") and emp._department:
            result["department"] = {"id": emp._department.id, "name": emp._department.name}
        if hasattr(emp, "_designation") and emp._designation:
            result["designation"] = {"id": emp._designation.id, "name": emp._designation.name}
        if hasattr(emp, "_entity") and emp._entity:
            result["entity"] = {
                "id": emp._entity.id,
                "tradeName": getattr(emp._entity, "trade_name", None),
                "legalName": getattr(emp._entity, "legal_name", None),
            }
        if hasattr(emp, "_location") and emp._location:
            result["location"] = {"id": emp._location.id, "name": emp._location.name}
        return result

    async def list(self, tenant_id: str, page: int = 1, page_size: int = 50, **filters) -> dict:
        skip = (page - 1) * page_size
        items, total = await self.repo.find_many(tenant_id, skip=skip, limit=page_size, **filters)
        return {
            "items": [self._serialize_with_relations(e) for e in items],
            "total": total,
            "page": page,
            "page_size": page_size,
            "total_pages": (total + page_size - 1) // page_size,
            "has_next": page * page_size < total,
            "has_prev": page > 1,
        }

    async def get_by_id(self, employee_id: str, tenant_id: str) -> dict | None:
        emp = await self.repo.find_by_id(employee_id, tenant_id)
        if not emp:
            return None
        return self._serialize_with_relations(emp)

    async def create(self, data: dict, tenant_id: str, actor_name: str = "System") -> dict:
        employee_code = data.get("employee_code", "").strip()
        first_name = data.get("first_name", "").strip()
        if not employee_code:
            raise HTTPException(status_code=400, detail="Employee code is required")
        if not first_name:
            raise HTTPException(status_code=400, detail="First name is required")

        existing = await self.repo.find_by_code(tenant_id, employee_code)
        if existing:
            raise HTTPException(status_code=409, detail="Employee code already exists")

        extras = {k: v for k, v in data.items() if k not in EMPLOYEE_FIELDS}
        custom_data = json.dumps(extras) if extras else None

        clean_data = {k: v for k, v in data.items() if k in EMPLOYEE_FIELDS}
        clean_data["tenant_id"] = tenant_id
        clean_data["custom_data"] = custom_data

        emp = await self.repo.create(clean_data)

        self.db.add(EmployeeTimelineEvent(
            tenant_id=tenant_id, employee_id=emp.id,
            event_type="Created", title="Employee record created",
            event_date=datetime.utcnow(), actor_name=actor_name,
        ))
        await self.db.commit()

        return self._serialize_with_relations(emp)

    async def update(self, employee_id: str, data: dict, tenant_id: str, actor_name: str = "System") -> dict:
        existing = await self.repo.find_by_id(employee_id, tenant_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Employee not found")

        if "employee_code" in data:
            new_code = data["employee_code"].strip()
            if not new_code:
                raise HTTPException(status_code=400, detail="Employee code cannot be empty")
            if new_code != existing.employee_code:
                clash = await self.repo.find_by_code(tenant_id, new_code)
                if clash:
                    raise HTTPException(status_code=409, detail="Employee code already exists")

        update_data = {}
        changed = []
        for field in EMPLOYEE_FIELDS:
            if field in data:
                old_val = getattr(existing, field, None)
                new_val = data[field]
                if old_val != new_val:
                    update_data[field] = new_val
                    changed.append({
                        "field": field,
                        "old": str(old_val) if old_val is not None else "",
                        "new": str(new_val) if new_val is not None else "",
                    })

        extras = {k: v for k, v in data.items() if k not in EMPLOYEE_FIELDS}
        if extras:
            update_data["custom_data"] = json.dumps(extras)

        if not update_data:
            return existing

        old_status = existing.employee_status
        new_status = update_data.get("employee_status", old_status)

        updated = await self.repo.update(employee_id, {**update_data, "tenant_id": tenant_id})

        if changed:
            for c in changed:
                self.db.add(EmployeeAuditLog(
                    tenant_id=tenant_id, employee_id=employee_id,
                    module="Personal", field=c["field"],
                    old_value=c["old"], new_value=c["new"],
                    action="Update", changed_by=actor_name,
                ))

        if new_status != old_status:
            self.db.add(EmployeeStatusHistory(
                tenant_id=tenant_id, employee_id=employee_id,
                old_status=old_status, new_status=new_status,
                effective_date=datetime.utcnow(), changed_by=actor_name,
            ))
            self.db.add(EmployeeTimelineEvent(
                tenant_id=tenant_id, employee_id=employee_id,
                event_type="Status changed",
                title=f'Status changed from "{old_status}" to "{new_status}"',
                event_date=datetime.utcnow(), actor_name=actor_name,
            ))

        await self.db.commit()
        return self._serialize_with_relations(updated)

    async def delete(self, employee_id: str, tenant_id: str) -> dict:
        existing = await self.repo.find_by_id(employee_id, tenant_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Employee not found")
        await self.repo.delete(employee_id)
        return {"ok": True}
